Test OOP Review

AdiMarianMutu opened the pull request.

AdiMarianMutu committed:
Create CoffeeShop.js

AdiMarianMutu committed:
Jasmine Unit Test Framework

AdiMarianMutu committed:
Update Readme.md

pirvudoru commented on CoffeeShop.js:
```diff
@@ -0,0 +1,106 @@
+		return CalculateDistance(user_xcoo, user_ycoo, a[1], a[2]) - CalculateDistance(user_xcoo, user_ycoo, b[1], b[2]);
+	});
+
+	// Retrieves the individual distance in km for the nearest coffee shop
```
How to write useful comments: 
https://blog.submain.com/field-guide-code-comments/

Usually you need to write self-documenting code. Nobody want s to read the same thing twice written differently (natural language / code)

Xxshark888xX replied:
> Requirements are not met. Please see comments.

I've fixed the formula which is used to retrieve the distance between two points on a flat surface

pirvudoru commented on CoffeeShop.js:
```diff
@@ -0,0 +1,106 @@
+	process.argv.forEach(function(val, index, array) {
+		switch(index) {
+			case 2:
+				user_xcoo = parseFloat(val);
```
https://www.smashingmagazine.com/2012/10/why-coding-style-matters/

Every language and project has a coding convention. Example: https://github.com/airbnb/javascript
This can be enforced with linting tools (eslint)

AdiMarianMutu replied:
Thank you for sharing this information with me, I'll read everything in order to improve my skills

pirvudoru commented on CoffeeShop.js:
```diff
@@ -0,0 +1,106 @@
+          c(x * p) * c(x1 * p) * 
+          (1 - c((y1 - y) * p)) / 2;
+
+  return 12742 * Math.asin(Math.sqrt(a)); // 2 * R; R = 6371 km
```
From the readme:
```
Assume all coordinates lie on a plane.
```

AdiMarianMutu replied:
I've fixed the formula which is used to retrieve the distance between two points on a flat surface

pirvudoru commented on CoffeeShop.js:
```diff
@@ -0,0 +1,106 @@
+	}
+}
+
+function ParseCSV(filePath) {
```
This looks more like FP than OOP.

pirvudoru commented on jasmine/spec/CoffeeShop_unit_test.js:
```diff
@@ -0,0 +1,31 @@
+        });
+
+		it("Should return '216.2894' km distance", function() {
+            expect(CalculateDistance(45.783852, 24.151526, 46.827218, 26.528130).toFixed(4)).toBeCloseTo(216.2894);
```
Why do specs test a function defined in specs only? Purpose for tests:

https://blog.hubstaff.com/why-you-should-write-unit-tests/
https://www.softwaretestingtricks.com/2007/01/unit-testing-versus-functional-tests.html

AdiMarianMutu replied:
> Why do specs test a function defined in specs only? Purpose for tests:
> 
> https://blog.hubstaff.com/why-you-should-write-unit-tests/
> https://www.softwaretestingtricks.com/2007/01/unit-testing-versus-functional-tests.html

I'm still new to Unit Tests.
Let me know if the new unit tests I wrote are doing better

pirvudoru commented on jasmine/SpecRunner.html:
```diff
@@ -0,0 +1,24 @@
+  <script src="../CoffeeShop_unit_test.js"></script>
+
+  <!-- include spec files here... -->
+  <script src="spec/CoffeeShop_unit_test.js"></script>
```
Every language has now at least one package manager. JS's most comonly used is https://www.npmjs.com/

To avoid adding library files to the main repository and make it more easy to update to newer versions.

pirvudoru commented on CoffeeShop.js:
```diff
@@ -0,0 +1,106 @@
+		if (require("fs").existsSync(filePath))
+			return true;
+
+		return false;
```
https://medium.com/@ntgard/returning-boolean-literals-is-a-code-smell-7a39531d6b60

AdiMarianMutu replied:
Thanks for the article!

pirvudoru commented on CoffeeShop.js:
```diff
@@ -0,0 +1,106 @@
+	// # MAIN CODE # //
+
+	// Retrieves the first 3 nearest coffee shop
+	(shopCsv_fileData = ParseCSV(shopCsv_filePath)).sort(function(a, b) {
```
https://medium.com/coding-skills/clean-code-101-meaningful-names-and-functions-bf450456d90c

pirvudoru reviewed (changes requested):
Requirements are not met. Please see comments.

alexandru-calinoiu reviewed (changes requested):
Is a good practice to include in the description of the pr how to test and run it.

You should include at minimum:
- a description, rought outline of what it does
- system req (like node version)
- setup instructions (do we need to install packages? npm, yarn?)
- how to run the code (sample input if the case might be)
- how to run the tests

levelboy reviewed (changes requested):
PRO:
It works as you intended it to work

CONS:
Distance is calculated in a complicated way.
The solution is not Object Oriented at all. Even the class that you have there has all methods as static. 
There's a lot of code smells. Take a look here: https://blog.codinghorror.com/code-smells/ Can you go through the list and see how many you identify. Post the list with all the code smells here. Also, try to fix them.

AdiMarianMutu committed:
Update Readme.md

Updated the 'How to run the code' section

AdiMarianMutu committed:
Update CoffeeShop.js

 * Fixed the formula used to calculate the distance between the user and the coffee shop
 * General code refactoring
 * Improved the unit tests (now are console based)

AdiMarianMutu committed:
Obsolete

AdiMarianMutu committed:
obsolete

AdiMarianMutu committed:
obsolete

AdiMarianMutu committed:
obsolete

AdiMarianMutu committed:
obsolete

AdiMarianMutu committed:
obsolete

AdiMarianMutu committed:
obsolete

AdiMarianMutu committed:
Unit Test

Improved the unit tests.

Please read the [readme.md](https://github.com/Xxshark888xX/test_oop/blob/master/Readme.md) for the execution instructions

AdiMarianMutu commented:
> Is a good practice to include in the description of the pr how to test and run it.
> 
> You should include at minimum:
> 
> * a description, rought outline of what it does
> * system req (like node version)
> * setup instructions (do we need to install packages? npm, yarn?)
> * how to run the code (sample input if the case might be)
> * how to run the tests

Please read the 'readme.md' file.
You'll find everything you need to know in that file.

AdiMarianMutu committed:
Changed the 'toBeCloseTo' method to 'toBe'

alexandru-calinoiu commented:
Mersi @Xxshark888xX o sa revenim in scurt timp la tine cu un email.
